app/src/main/kotlin/com/mapzen/erasermap/view/MultiModalDirectionListAdapter.kt

```diff
@@ -0,0 +1,423 @@
+      }
+    } else {
+      holder.prevTransitLine.visibility = View.GONE
+    }
+
+    for (i in 0..holder.stationNamesContainer.childCount - 1) {
+      recycler.queueView(holder.stationNamesContainer.getChildAt(i), VIEW_TYPE_TRANSIT_ROW)
+    }
+    holder.stationNamesContainer.removeAllViews()
+    val transitStops = instruction.getTransitInfo()?.getTransitStops() as ArrayList<TransitStop>
+
+    // we show all stops except first and last which are shown in larger font
+    var numStops = transitStops.size - 2
+    if (listItem.nextItemMode != null) {
+      if (listItem.nextItemMode == TravelMode.PEDESTRIAN) {
+        holder.endingStationDot.visibility = View.VISIBLE
+        holder.endingStationName.visibility = View.VISIBLE
+        holder.pedestrianConnector.visibility = View.VISIBLE
+      } else {
+        numStops++
+        holder.endingStationDot.visibility = View.GONE
+        holder.endingStationName.visibility = View.GONE
+        holder.pedestrianConnector.visibility = View.GONE
+      }
+    }
+    if (listItem.expanded) {
+
+      for (i in 1..numStops) {
+        var stationRow = recycler.dequeueView(VIEW_TYPE_TRANSIT_ROW)
+        if (stationRow == null) {
+          stationRow = View.inflate(context, R.layout.transit_station_row, null)
+        }
+        val stationName = stationRow?.findViewById(R.id.station_name) as TextView
+        stationName.text = transitStops[i].getName()
+        holder.stationNamesContainer.addView(stationRow)
+      }
+    }
+    holder.endingStationName.text = transitStops[transitStops.size-1].getName()
+  }
+
+  open class ViewHolder() {
+  }
+
+  class IconTitleViewHolder(view: View): ViewHolder() {
+    val titleView: TextView
+    val iconView: ImageView
+
+    init {
+      titleView = view.findViewById(R.id.title) as TextView
+      iconView = view.findViewById(R.id.image_view) as ImageView
+    }
+  }
+
+  class PedestrianViewHolder(view: View, context: Context): ViewHolder() {
+    val totalDistanceView: DistanceView
+    val totalTimeView: TimeView
+    val instructionsContainer: LinearLayout
+    val dashedLine: View
+    val distanceTimeContainer: DistanceTimeExpanderLayout
+
+    init {
+      totalDistanceView = view.findViewById(R.id.total_distance) as DistanceView
+      totalTimeView = view.findViewById(R.id.total_time) as TimeView
+      instructionsContainer = view.findViewById(R.id.instructions_container) as LinearLayout
+      dashedLine = view.findViewById(R.id.dashed_line)
+      distanceTimeContainer = view.findViewById(R.id.distance_time_container)
+          as DistanceTimeExpanderLayout
+      val textView = view.findViewById(R.id.distance_time_text_view) as TextView
+      val builder = StringBuilder()
+      builder.append(context.getString(R.string.comma))
+      builder.append(" ")
+      textView.text = builder.toString()
+    }
+  }
+
+  class TransitViewHolder(view: View, context: Context): ViewHolder() {
```

Comment:
They follow the "view holder" pattern and allow access to the underlying view directly, without having to call `findViewByID` everytime they are accessed